src/client_side.h

```diff
@@ -199,6 +199,26 @@ class ConnStateData : public Server, public HttpControlMsgSink, private Independ
     // pining related comm callbacks
     virtual void clientPinnedConnectionClosed(const CommCloseCbParams &io);
 
+    /// noteTakeServerConnectionControl() callback parameter
+    class ServerConnectionContext {
+    public:
+        ServerConnectionContext(const Comm::ConnectionPointer &conn, const HttpRequest::Pointer &req, const SBuf &post101Bytes): request(req), preReadServerBytes(post101Bytes), conn_(conn) { conn_->enterOrphanage(); }
+
+        /// gives to-server connection to the new owner
+        Comm::ConnectionPointer connection() { conn_->leaveOrphanage(); return conn_; }
+
+        HttpRequest::Pointer request; ///< the last to-server request on connection
```

Comment:
Yes. It is actually only used to check that we are referring to the correct request (See the "assert(http->request == server.request.getRaw());" inside Http::One::Server::noteTakeServerConnectionControl).

Reply:
> I am still considering whether to merge HttpControlMsg and ServerConnectionContext to avoid the extra exchange between the two sides of the transaction.

You need to store server info to ConnStateData (or Http::One::Server) until you get confirmation that the control Msg written to the client. These classes are already hold a lot of info. In the other hand you will save two async calls on each upgraded connection.

The question is two extra async calls for each upgraded connection vs some (one or two) extra Http::OneServer class members and some extra bytes (just a pointer?) for each HTTP connection.

Reply:
Please remove this data member.

Reply:
> Please remove this data member.
commit f0f86006